core/metacling/src/TClingCallbacks.cxx

```diff
@@ -362,6 +362,16 @@ bool TClingCallbacks::LookupObject(const DeclContext* DC, DeclarationName Name)
    if (Name.getNameKind() != DeclarationName::Identifier)
       return false;
 
+   Sema &SemaR = m_Interpreter->getSema();
+   auto *D = dyn_cast<Decl>(DC);
+   if (D) {
+      SourceLocation Loc = D->getLocation();
+      if (Loc.isValid() && SemaR.getSourceManager().isInSystemHeader(Loc)) {
```

Comment:
We could ask if `Decl`s `OwningModule` is a system one. This way we could "annotate" modules which we do not need to mess with. We consider modules starting with `ROOT_` system modules because a pure `[system]` tag also changes a little the semantics of the whole thing but I think we are fine to do that check here.

I think we probably can explore a little more the heuristics you came up with as it looks promising. Namely, can a declaration context of a module be 'extended with' declarations from a module that is not listed among the module dependencies? I suspect the only cases where this can happen is friend declarations and ADL lookups. I don't think we need to support that... Eg. what happens if we have a module that adds another `std::swap` with `MyClass` parameter - are we expected to autoload the module exporting `MyClass`?

Reply:
I like asking the `OwningModule` and changed both the existing and the new check :smiley:

> I think we probably can explore a little more the heuristics you came up with as it looks promising.

Yes, we likely need to understand the other failure modes and then abstract that to a heuristic / check that we can add here. For now I just want to make both `LookupObject` methods bail out on the same conditions, namely a `Decl` coming from a system module.

Reply:
Hm well, I guess this doesn't work for the `nortcxxmod` case? :grimacing:

Reply:
I don't think we care for the `nortcxxmod` as the PCH is preloaded and we don't react to entities that come from an `ASTFile`. I don't think we have "system" headers which are not in the PCH.

Reply:
I reverted back to asking `isInSystemHeader` which did better in the tests...